File: backend/app/routers/discussion.py
from fastapi import APIRouter, Depends, status

from app.core.dependencies import get_current_user
from app.db.connection import get_db
from app.schemas.discussion import (
    CreateCommentInput,
    CreatePostInput,
    UpdateCommentInput,
    UpdatePostInput,
)
from app.services import discussion_service

router = APIRouter(tags=["Discussions"])


# ── Teacher: manage posts ─────────────────────────────────────────────────────

@router.post("/classrooms/{classroom_id}/discussions", status_code=status.HTTP_201_CREATED)
def create_post(
    classroom_id: int,
    body: CreatePostInput,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.create_post(
        conn,
        classroom_id=classroom_id,
        created_by=current_user["id"],
        title=body.title,
        content=body.content,
    )


@router.put("/classrooms/{classroom_id}/discussions/{post_id}")
def update_post(
    classroom_id: int,
    post_id: int,
    body: UpdatePostInput,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.update_post(
        conn,
        classroom_id=classroom_id,
        post_id=post_id,
        owner_id=current_user["id"],
        title=body.title,
        content=body.content,
        is_active=body.is_active,
    )


@router.delete("/classrooms/{classroom_id}/discussions/{post_id}")
def delete_post(
    classroom_id: int,
    post_id: int,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.delete_post(conn, classroom_id, post_id, current_user["id"])


# ── Anyone: view posts ────────────────────────────────────────────────────────

@router.get("/classrooms/{classroom_id}/discussions")
def get_posts(
    classroom_id: int,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.get_posts(conn, classroom_id, current_user["id"])


@router.get("/classrooms/{classroom_id}/discussions/{post_id}")
def get_post(
    classroom_id: int,
    post_id: int,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.get_post(conn, classroom_id, post_id, current_user["id"])


# ── Anyone: comments and replies ─────────────────────────────────────────────

@router.post("/classrooms/{classroom_id}/discussions/{post_id}/comments", status_code=status.HTTP_201_CREATED)
def create_comment(
    classroom_id: int,
    post_id: int,
    body: CreateCommentInput,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.create_comment(
        conn,
        classroom_id=classroom_id,
        post_id=post_id,
        user_id=current_user["id"],
        content=body.content,
        parent_id=body.parent_id,
    )


@router.put("/classrooms/{classroom_id}/discussions/{post_id}/comments/{comment_id}")
def update_comment(
    classroom_id: int,
    post_id: int,
    comment_id: int,
    body: UpdateCommentInput,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.update_comment(
        conn,
        classroom_id=classroom_id,
        comment_id=comment_id,
        user_id=current_user["id"],
        content=body.content,
    )


@router.delete("/classrooms/{classroom_id}/discussions/{post_id}/comments/{comment_id}")
def delete_comment(
    classroom_id: int,
    post_id: int,
    comment_id: int,
    current_user: dict = Depends(get_current_user),
    conn=Depends(get_db),
):
    return discussion_service.delete_comment(
        conn,
        classroom_id=classroom_id,
        post_id=post_id,
        comment_id=comment_id,
        user_id=current_user["id"],
    )

File: frontend/src/actions/dashboard/discussion.ts
// actions/dashboard/discussion.ts

import { private_api_call } from "../private_api_call";



// import {
//     getPosts,
//     getPost,
//     createPost,
//     updatePost,
//     deletePost,
//     createComment,
//     updateComment,
//     deleteComment,
// } from "@/actions/dashboard/discussion";

export async function createPost(
    classroomId: string,
    postTitle: string,
    postContent: string
) {
    const res = await private_api_call({
        method: "POST",
        path: `classrooms/${classroomId}/discussions`,
        body: {
            title: postTitle,
            content: postContent || null,
        },
    });

    return res;
}

export async function updatePost(classroomId: string, postId: number, postTitle: string, postContent: string, isActive: boolean) {

    const res = await private_api_call({
        method: "PUT",
        path: `classrooms/${classroomId}/discussions/${postId}`,
        body: {
            title: postTitle,
            content: postContent || null,
            is_active: isActive,
        },
    });
    return res;


}

export async function deletePost(classroomId: string, postId: number) {

    const res = await private_api_call({
        method: "DELETE",
        path: `classrooms/${classroomId}/discussions/${postId}`,
    });
    return res;
}

export async function getPost(classroomId: string, postId: number) {

    const res = await private_api_call({
        method: "GET",
        path: `classrooms/${classroomId}/discussions/${postId}`,
    });
    return res;

}

export async function addComment(classroomId: string, postId: number, parentId: number | null, content: string) {

    const res = await private_api_call({
        method: "POST",
        path: `classrooms/${classroomId}/discussions/${postId}/comments`,
        body: {
            content,
            parent_id: parentId,
        },
    });

    return res;
}

export async function updateComment(classroomId: string, postId: number, commentId: number, content: string) {


    const res = await private_api_call({
        method: "PUT",
        path: `classrooms/${classroomId}/discussions/${postId}/comments/${commentId}`,
        body: {
            content,
        },
    });
    return res;
}

export async function getPosts(classroomId: string) {
    const res = await private_api_call({
        method: "GET",
        path: `classrooms/${classroomId}/discussions`,
    });
    return res;

}

export async function deleteComment(classroomId: string, postId: number, commentId: number) {


    const res = await private_api_call({
        method: "DELETE",
        path: `classrooms/${classroomId}/discussions/${postId}/comments/${commentId}`,
    });
    return res;

}

